Add JDK17 and springboot 3.0 support

skyao opened the pull request:
# Description

Upgrade to add springboot 3.0 support in dapr java sdk.

For the details, please reference to the proposal: https://github.com/dapr/java-sdk/issues/971 

## Issue reference

https://github.com/dapr/java-sdk/issues/971

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

mukundansundar commented on .github/workflows/build.yml:
```diff
@@ -21,22 +21,19 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        java: [ 11, 13, 15, 16 ]
```
Do not remove 11 as yet. Have that also

skyao replied:
I reference to the previous design as:

```toml
      matrix:
        java: [ 11, 13, 15, 16 ]
        spring-boot-version: [ 2.7.8 ]
        experimental: [ false ]
        include:
          - java: 11
            spring-boot-version: 2.6.14
            experimental: false
          - java: 11
            spring-boot-version: 2.5.7
            experimental: false
          - java: 11
            spring-boot-version: 2.4.0
            experimental: false
          - java: 11
            spring-boot-version: 2.3.6.RELEASE
            experimental: false
```

Through java8 is supported, sdk-tests subproject used jdk11 and set compile source and target to jdk11, and test with springboot 2.x versions with JDK11 as list above.

skyao replied:
If a jdk version below jdk17 appears here, it means that no jdk17 features can be used in the sdk-tests sub-project, or it will fail to compile.

This means that the sdk-tests project has to be split into two sub-projects, jdk17 and jdk11, and the various test cases have to be rewritten in both sub-projects. Otherwise it would not be covered by both jdk17 and jdk11.

skyao replied:
That's why it costs me so much to keep the sdk-tests subproject from being split.

mukundansundar replied:
@skyao For now we are not using any jdk17 features in the idk-tests sub-project right? 
And as long as we support jdk8 and jdk11, won't that continue?

mukundansundar commented on sdk-tests/pom.xml:
```diff
@@ -12,16 +12,16 @@
-    <maven.compiler.source>11</maven.compiler.source>
-    <maven.compiler.target>11</maven.compiler.target>
+    <maven.compiler.source>17</maven.compiler.source>
+    <maven.compiler.target>17</maven.compiler.target>
```
why is the target changing here?

mukundansundar commented on examples/pom.xml:
```diff
@@ -18,12 +18,13 @@
-    <java.version>11</java.version>
+    <java.version>17</java.version>
     <maven.compiler.source>${java.version}</maven.compiler.source>
     <maven.compiler.target>${java.version}</maven.compiler.target>
```
can the target still be 8/11? why does that need to change?

skyao replied:
As we discussed, we will use jdk17 and springboot3 for examples and sdk-tests subprojects.

mukundansundar commented on spotbugs-exclude.xml:
```diff
@@ -3,4 +3,34 @@
         <Package name="~io\.dapr\.examples.*"/>
     </Match>
+
+    <!--Temporarily ignoring checking after upgrade to new spotbugs version-->
+    <Match>
+        <Package name="~io\.dapr.*"/>
+        <Bug pattern="EI_EXPOSE_REP"/>
+    </Match>
+
+    <!--Temporarily ignoring checking after upgrade to new spotbugs version-->
+    <Match>
+        <Package name="~io\.dapr.*"/>
+        <Bug pattern="CT_CONSTRUCTOR_THROW"/>
+    </Match>
+
+    <!--Temporarily ignoring checking after upgrade to new spotbugs version-->
+    <Match>
+        <Package name="~io\.dapr.*"/>
+        <Bug pattern="SS_SHOULD_BE_STATIC"/>
+    </Match>
+
+    <!--Temporarily ignoring checking after upgrade to new spotbugs version-->
+    <Match>
+        <Package name="~io\.dapr.*"/>
+        <Bug pattern="EI_EXPOSE_REP"/>
+    </Match>
+
+    <!--Temporarily ignoring checking after upgrade to new spotbugs version-->
+    <Match>
+        <Package name="~io\.dapr.*"/>
+        <Bug pattern="EI_EXPOSE_REP2"/>
+    </Match>
```
is there a reason why we need this?

skyao replied:
https://github.com/dapr/java-sdk/issues/971

See “Spotbugs bug pattern” section in Appendices  part.

mukundansundar reviewed (dismissed):
lgtm. waiting for tests to pass

artursouza commented on sdk-springboot30/pom.xml:
```diff
@@ -0,0 +1,137 @@
+    <version>1.11.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>dapr-sdk-springboot30</artifactId>
```
Just dapr-sdk-springboot3, please.

skyao replied:
OK. updated!

artursouza commented on sdk-tests/src/test/java/io/dapr/it/tracing/http/OpenTelemetryInterceptor.java:
```diff
@@ -16,15 +16,14 @@
-import javax.servlet.DispatcherType;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
+import jakarta.servlet.DispatcherType;
```
Why not javax? Is it required because of Java17?

skyao replied:
Yes, Java 17 removed `javax.*` and required to use `jakarta.*`

artursouza commented:
Validation of examples is failing.

skyao commented:
I have confused by the behavior of NOP Logger and backlog. 

Before this PR, examples subproject in java-sdk repo depends on springboot2.x and introduced logback as its slf4j implementation:

```bash
[INFO] +- org.springframework.boot:spring-boot-starter-web:jar:2.7.8:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter:jar:2.7.8:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-starter-logging:jar:2.7.8:compile
[INFO] |  |  |  +- ch.qos.logback:logback-classic:jar:1.2.11:compile
[INFO] |  |  |  |  \- ch.qos.logback:logback-core:jar:1.2.11:compile
```

In this PR, I updated the version of springboot to springboot 3.0, also introduced logback as its slf4j implementation:

```bash
[INFO] +- org.springframework.boot:spring-boot-starter-web:jar:3.0.13:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter:jar:3.0.13:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-starter-logging:jar:3.0.13:compile
[INFO] |  |  |  +- ch.qos.logback:logback-classic:jar:1.4.11:compile
[INFO] |  |  |  |  \- ch.qos.logback:logback-core:jar:1.4.11:compile
```

But now the github action failed because of the conflicts between logback and NOP logger from slf4j-api: 

```bash
		== APP == 	... 8 more
		== APP == Caused by: java.lang.IllegalArgumentException: LoggerFactory is not a Logback LoggerContext but Logback is on the classpath. Either remove Logback or the competing implementation (class org.slf4j.helpers.NOPLoggerFactory loaded from jar:file:/home/runner/work/java-sdk/java-sdk/examples/target/dapr-java-sdk-examples-exec.jar!/BOOT-INF/lib/slf4j-api-1.7.25.jar!/). If you are using WebLogic you will need to add 'org.slf4j' to prefer-application-packages in WEB-INF/weblogic.xml: org.slf4j.helpers.NOPLoggerFactory
```

We used NOP logger in workflow, in DaprWorkflowContextImpl: 

```java
  public Logger getLogger() {
    if (this.innerContext.getIsReplaying()) {
      return NOPLogger.NOP_LOGGER;
    }
    return this.logger;
  }
```

The problem is that it works with jdk11 + springboot2.x (before this PR), but it doesn't work with jdk17 + springboot3.0. 

I tried to exclude logback from examples subproject to fix above conflicts, then all the examples will pass exception workflow example: withtout logback, only NOP logger it can't output the text for validation.

```bash
== APP == SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
== APP == SLF4J: Defaulting to no-operation (NOP) logger implementation
== APP == SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

I also checked other examples to see why they didn't fail without logback, there are 

- using system out
- or using java.util.logging.Logger

skyao commented:
Done!

```java
   <dependency>
        <groupId>org.slf4j</groupId>
        <artifactId>slf4j-api</artifactId>
        <version>2.0.9</version>
      </dependency>
```

After adding slf4j-api  as compile dependency and upgrade its version to 2.0.9, then logback and NOP logger can work together.

@artursouza @mukundansundar This PR can be merged now!
